gems/aws-sdk-s3/lib/aws-sdk-s3/plugins/http_200_errors.rb

```diff
@@ -27,6 +27,27 @@ def check_for_error(context)
               error_code = xml.match(/<Code>(.+?)<\/Code>/)[1]
               error_message = xml.match(/<Message>(.+?)<\/Message>/)[1]
               S3::Errors.error_class(error_code).new(context, error_message)
+            elsif incomplete_response?(xml, context)
+              Seahorse::Client::NetworkingError.new(
+                S3::Errors
+                  .error_class('InternalError')
+                  .new(context, 'Empty response body')
+              )
+            end
+          end
+
+          private
+
+          # An S3 response is considered complete only if all output shapes
+          # expected in the body are present
+          def incomplete_response?(xml, context)
```

Comment:
I don't believe it is necessary to search the body for each of the expected top-level members of the document. If the XML document is opened but not closed correctly it would fail to parse. A minor simplification might be to check only for the document closing element.

Reply:
I've updated the PR with a similar approach - checking for the start of a valid xml tag.  If an xml tag is started but the body is incomplete, all of the XML engines will raise an xml ParsingError (except Oga, in the case of an incomplete closing tag.  see: https://gitlab.com/yorickpeterse/oga/-/issues/203).